src/remote.ts

```diff
@@ -0,0 +1,178 @@
+
+    await this.ensureClientDir()
+    const output = path.join(this.clientRoot, this.updatedVersion)
+
+    if (!await fs.pathExists(output)) {
+      await this.initializeDownload(output, manifest)
+    }
+
+    await this.setChannel()
+    await this.createBin(this.updatedVersion)
+    await this.touch()
+    await this.reexec()
+  }
+
+  private async determineCurrentVersion(): Promise<string|undefined> {
+    try {
+      const currentVersion = await fs.readFile(this.clientBin, 'utf8')
+      const matches = currentVersion.match(/\.\.[/|\\](.+)[/|\\]bin/)
+      return matches ? matches[1] : this.config.version
+    } catch (error) {
+      this.debug(error)
+    }
+    return this.config.version
+  }
+
+  protected async downloadAndExtract(output: string, gzUrl: string, baseDir: string, sha256gz?: string) {
+    const filesize = (n: number): string => {
+      const [num, suffix] = require('filesize')(n, {output: 'array'})
+      return num.toFixed(1) + ` ${suffix}`
+    }
+
+    const http: typeof HTTP = require('http-call').HTTP
+    const {response: stream} = await http.stream(gzUrl, this.getReqHeaders())
+    stream.pause()
+
+    const extraction = extract(stream, baseDir, output, sha256gz)
+
+    // to-do: use cli.action.type
+    if ((cli.action as any).frames) {
+      // if spinner action
+      const total = parseInt(stream.headers['content-length']!, 10)
+      let current = 0
+      const updateStatus = throttle(
+        (newStatus: string) => {
+          cli.action.status = newStatus
+        },
+        250,
+        {leading: true, trailing: false},
+      )
+      stream.on('data', data => {
+        current += data.length
+        updateStatus(`${filesize(current)}/${filesize(total)}`)
+      })
+    }
+
+    stream.resume()
+    await extraction
+  }
+
+  private async logChop() {
+    try {
+      this.debug('log chop')
+      const logChopper = require('log-chopper').default
+      await logChopper.chop(this.config.errlog)
+    } catch (error) {
+      this.debug(error.message)
+    }
+  }
+
+  private async reexec() {
+    cli.action.stop()
+    return new Promise((_, reject) => {
+      this.debug('restarting CLI after update', this.clientBin)
+      spawn(this.clientBin, ['update'], {
+        stdio: 'inherit',
+        env: {...process.env, [this.config.scopedEnvVarKey('HIDE_UPDATED_MESSAGE')]: '1'},
+      })
+      .on('error', reject)
+      .on('close', (status: number) => {
+        try {
+          if (status > 0) this.exit(status)
+        } catch (error) {
+          reject(error)
+        }
+      })
+    })
+  }
+
+  private async setChannel() {
+    const channelPath = path.join(this.config.dataDir, 'channel')
+    fs.writeFile(channelPath, this.channel, 'utf8')
+  }
+
+  private async skipUpdate(): Promise<string | false> {
+    if (!this.config.binPath) {
```

Comment:
I keep running into this when testing. Not clear on how / why this should be set. Seems to normally come from an env var? I also see `binPath` was added here: https://github.com/oclif/config/commit/f1d5a7f08c70a4c9531a2d12f73d548a40d6274b